File: cleanup.rs
use std::error::Error;

type Try<Value> = Result<Value, Box<dyn Error>>;

#[allow(dead_code)]
fn make_error<Value>(error: &str) -> Try<Value> {
    use std::io::{Error as IoError, ErrorKind};
    Err(IoError::new(ErrorKind::Other, error).into())
}

struct Resource {
    name: String,
}

impl Resource {
    fn new(name: &str) -> Try<Resource> {
        println!("open {}", name);
        Ok(Resource { name: name.into() })
    }
}

impl Drop for Resource {
    fn drop(&mut self) {
        println!("close {}", self.name);
    }
}

fn prep_out(out_name: &str, prep_names: &[&str]) -> Try<Resource> {
    let writer = Resource::new(out_name)?;
    for name in prep_names {
        let _reader = Resource::new(name)?;
        // if true {
        //     return make_error("WELCOME TO YOUR DOOM!!");
        // }
        println!("use {}", name);
    }
    Ok(writer)
}

fn main() -> Try<()> {
    let _writer = prep_out("out", &["a", "b"])?;
    println!("use out");
    Ok(())
}

File: cleanup.lua
Resource = {}

function Resource:open(name)
    print(string.format("open %s", name))
    local result = {name = name}
    setmetatable(result, self)
    self.__index = self
    return result
end

function Resource:close()
    print(string.format("close %s", self.name))
end

function Resource:__close(err)
    -- print(err)
    self:close()
end

function prep_out(out_name, prep_names)
    local writer = Resource:open("out")
    local ok, err = pcall(function()
        for _, name in ipairs(prep_names) do
            local reader <close> = Resource:open(name)
            -- error("WELCOME TO YOUR DOOM!!")
            print(string.format("use %s", name))
        end
    end)
    if ok then
        return writer
    else
        writer:close()
        error(err, 0)
    end
end

function main()
    local writer <close> = prep_out("out", {"a", "b"})
    print("use out")
end

main()
